src/printer.js

```diff
@@ -621,7 +621,40 @@ function genericPrintNoParens(path, options, print) {
         } else {
           parts.push(printPropertyKey(path, print));
         }
-        parts.push(": ", path.call(print, "value"));
+        let noBreak = [
+          "ArrayExpression",
+          "ArrayPattern",
+          "ArrowFunctionExpression",
+          "AssignmentPattern",
+          "CallExpression",
+          "FunctionExpression",
+          "Identifier",
+          "Literal",
+          "LogicalExpression",
+          "MemberExpression",
+          "NewExpression",
+          "ObjectExpression",
+          "ObjectPattern",
+          "StringLiteral",
+          "ThisExpression",
+          "TypeCastExpression"
+        ].reduce((stack, type) => stack || (n.value.type === type), false);
+        let printedValue = path.call(print, "value");
+        parts.push(
+          group(
+            concat([
+              multilineGroup(
+                concat([
+                  noBreak ? ": " : ": (",
+                  noBreak
+                    ? printedValue
+                    : indent(options.tabWidth, concat([line, printedValue]))
+                ])
+              ),
+              noBreak ? "" : concat([line, ")"])
```

Comment:
There are many different cases for `noBreak` that it'd probably be clearer to just separate it out completely. So do:

```js
if(noBreak) {
  parts.push(group(concat([ ... ]))
}
else {
  parts.push(group(concat([ ... ]))
}
```

Defining these separately might repeat one or two things, but makes it a lot more readable for what happens in that case.